Upgrading to Airflow 2.0 doc

vikramkoka opened the pull request:
Added guide for Upgrading to 2.0 
This is based heavily on the UPGRADING_to_2.0.md file. The intent here is to make user facing and available as part of the Airflow documentation. This is still a work in progress.

@ashb @kaxil @dimberman

vikramkoka committed:
Added guide for Upgrading to 2.0
This is based heavily on the UPGRADING_to_2.0.md file. The intent here is
to make user facing and available as part of the Airflow documentation.
This is still a work in progress.

vikramkoka committed:
Added guide for Upgrading to 2.0
This is based heavily on the UPGRADING_to_2.0.md file. The intent here is to have a user facing guide
which is available as part of the Airflow documentation. This is still a work in progress.

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+To minimize friction for users upgrading from Airflow 1.10 to Airflow 2.0 and beyond, a "bridge"
+release and final 1.10 version will be made available. Airflow 1.10.14 includes support for various critical features
+that make it easy for users to test DAGs and make sure they are as compatible with Airflow 2.0, so that the . We strongly recommend that all users upgrading to Airflow 2.0 first
```
Not finiished sentence.

vikramkoka replied:
Thanks @potiuk will fix.

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+    from airflow.operators.docker_operator import DockerOperator
+
+You would now run this command to import the provider:
```
```suggestion
You would now run this command to install the provider:
```

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+    from airflow.providers.docker.operators.docker import DockerOperator
+
+Pleaes note that the backport provider packages are just backports of the provider packages compatible with Airflow 2.0.
+Those provider packages are installed automatically when you install airflow with extras.
```
```suggestion
In Apache Airflow 2.0+, those provider packages are installed automatically when you install Airflow with extras.
Several of the providers (http, ftp, sqlite, imap) are installed automatically when you install Airflow even without extras.
```

kaxil replied:
```suggestion
Those provider packages are installed automatically when you install Airflow with extras.
```

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+automatically installs the ``apache-airflow-providers-docker`` package.
+But you can manage/upgrade remove provider packages separately from the airflow core.
+
```
```suggestion
You can read more about providers at :doc:`apache-airflow-providers:index`.
```

kaxil commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+Airflow 1.10 will be the last release series to support Python 2. Airflow 2.0.0 will require Python 3.6+.
+
+If you have a specific task that still requires Python 2 then you can use the PythonVirtualenvOperator for this.
```
```suggestion
If you have a specific task that still requires Python 2 then you can use the :class:`~airflow.operators.python.PythonVirtualenvOperator` for this.
```

dimberman replied:
```suggestion
If you have a specific task that still requires Python 2 then you can use the PythonVirtualenvOperator or the KubernetesPodOperator for this.
```

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+You don't need to make changes to the configuration file as the FAB RBAC UI is
+the only supported UI.
+
+.. code-block:: bash
```
Repeated code block (15 lines above). This one can be removed.

kaxil commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+To minimize friction for users upgrading from Airflow 1.10 to Airflow 2.0 and beyond, a "bridge"
+release and final 1.10 version will be made available. Airflow 1.10.14 includes support for various critical features
```
We need to convey something along the lines of what we mention in https://github.com/apache/airflow/blob/master/UPGRADING_TO_2.0.md#support-for-airflow-110x-releases

vikramkoka replied:
Good catch. I missed that, will add it.

vikramkoka replied:
Actually, it is here, but in the tail end of the document. Do you want me to move it up?

vikramkoka replied:
I pulled up some of the content from later on to address your feedback that it should be earlier. Does this work?

kaxil commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+     airflow generate_pod_template -o <output file path>
+
+Once you have performed this step, simply write out the file path to this file in the ``pod_template_file`` section of the ``kubernetes``
```
```suggestion
Once you have performed this step, simply write out the file path to this file in the ``pod_template_file`` config of the ``kubernetes``
```

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+Migration Guide from Experimental API to Stable API v1
+''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+In Airflow 2.0, we added the new REST API. Experimental API still works, but support may be dropped in the future.
```
```suggestion
In Airflow 2.0, we added the new REST API. Experimental API still works, but support may be dropped in the future.

The experimental API however does not require authentication, so it is disabled by default. You need to explicitly enable the experimental API if you want to use it.
```

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+In Airflow 2.0, we added the new REST API. Experimental API still works, but support may be dropped in the future.
+If your application is still using the experimental API, you should consider migrating to the stable API.
```
```suggestion
If your application is still using the experimental API, you should **seriously** consider migrating to the stable API.
```

kaxil commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+Step 4: Set Operators to Backport Providers
+'''''''''''''''''''''''''''''''''''''''''''
+
+Now that you are set up in airflow 1.10.14 with python a 3.6+ environment, you are ready to start porting your DAGs to Airfow 2.0 compliance!
```
```suggestion
Now that you are set up in Airflow 1.10.14 with Python a 3.6+ environment, you are ready to start porting your DAGs to Airflow 2.0 compliance!
```

kaxil commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+Now that you are set up in airflow 1.10.14 with python a 3.6+ environment, you are ready to start porting your DAGs to Airfow 2.0 compliance!
+
+The most important step in this transition is also the easiest step to do in pieces. All Airflow 2.0 operators are backwards compatible with Airflow 1.10
+using the [backport providers](./backport-providers.rst) service. In your own time, you can transition to using these backport-providers
```
```suggestion
using the [backport providers](./backport-providers.rst) packages. In your own time, you can transition to using these backport-providers
```

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+::
+    This endpoint ``/api/v1/dags/{dag_id}/dagRuns`` also allows you to filter dag_runs with parameters such as ``start_date``, ``end_date``, ``execution_date`` etc in the query string.
+    Therefore the operation previously performed by this endpoint
```
```suggestion
    Therefore the operation previously performed by this endpoint:
```

dimberman commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+Features in 1.10.14 include:
+
+1. Most breaking DAG and architecture changes of Airflow 2.0 have been backported to Airflow 1.10.14. This backward-compatibility does not mean
+that 1.10.14 will process these DAGs the same way as Airflow 2.0. What this does mean is that most Airflow 2.0
```
```suggestion
that 1.10.14 will process these DAGs the same way as Airflow 2.0. Instead, this means that most Airflow 2.0
```

kaxil commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+.. code-block:: bash
+
+    pip install airflow[docker]
```
```suggestion
    pip install 'apache-airflow[docker]'
```

kaxil commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+    pip install airflow[docker]
+
+automatically installs the ``apache-airflow-providers-docker`` package.
+But you can manage/upgrade remove provider packages separately from the airflow core.
```
```suggestion
But you can manage/upgrade/remove provider packages separately from the Airflow core.
```

vikramkoka committed:
Apply suggestions from code review

Incorporated feedback

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>
Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

vikramkoka committed:
Apply suggestions from code review

Fixed suggestions based on review

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

dimberman commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+*The KubernetesExecutor Will No Longer Read from the airflow.cfg for Base Pod Configurations.*
+
+In Airflow 2.0, the KubernetesExecutor will require a base pod template written in yaml. This file can exist
+anywhere on the host machine and will be linked using the ``pod_template_file`` configuration in the airflow.cfg.
```
```suggestion
anywhere on the host machine and will be linked using the ``pod_template_file`` configuration in the airflow.cfg. You can create a ``pod_template_file`` by running the following command:  ``airflow generate_pod_template`` ```

potiuk commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+After the Airflow 2.0 GA (General Availability) release, it expected that all
+future Airflow development would be based on Airflow 2.0, including a series of
+patch releases such as 2.0.1, 2.0.2 and then feature releases such as 2.1.
+
```
```suggestion

We plan to follow strict semver approach to our versioning, which means that we do not plan to make any backwards-incompatible changes in 2.* release series. Breaking changes (including deprecated features) will happen in 3.0.0 release. 
```

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+Step 3: Install and run the Upgrade check scripts
+'''''''''''''''''''''''''''''''''''''''''''''''''
+
+After upgrading to Airflow 1.10.14, we recommend that you install the "upgrade check" scripts. These scripts will read through your ``airflow.cfg`` and all of your Dags and will give a detailed report of all changes required before upgrading. We are testing this script diligently, and our goal is that any Airflow setup that can pass these tests will be able to upgrade to 2.0 without any issues.
```
```suggestion
After upgrading to Airflow 1.10.14, we recommend that you install the "upgrade check" scripts. These scripts will read through your ``airflow.cfg`` and all of your DAGs and will give a detailed report of all changes required before upgrading. We are testing this script diligently, and our goal is that any Airflow setup that can pass these tests will be able to upgrade to 2.0 without any issues.
```

vikramkoka committed:
Apply suggestions from code review

Incorporated suggestions from review

Co-authored-by: Daniel Imberman <daniel.imberman@gmail.com>
Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>
Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+More details about this process are here  :ref:`Upgrade Check Scripts<upgrade-check:upgrade_check>`
+
+
+Step 4: Set Operators to Backport Providers
```
```suggestion
Step 4: Import Operators from Backport Providers
```

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+Now that you are set up in Airflow 1.10.14 with Python a 3.6+ environment, you are ready to start porting your DAGs to Airflow 2.0 compliance!
+
+The most important step in this transition is also the easiest step to do in pieces. All Airflow 2.0 operators are backwards compatible with Airflow 1.10
+using the [backport providers](./backport-providers.rst) packages. In your own time, you can transition to using these backport-providers
```
Markdown link syntax left over here.

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+    pip install 'apache-airflow[docker]'
+
+automatically installs the ``apache-airflow-providers-docker`` package.
```
This install provider via extra only applies to Airflow 2, which isn't clear (to me) from this section.

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+rendering involving undefined variables will fail the task, as well as displaying an error in the UI when
+rendering.
+
+The behavior can be reverted when instantiating a DAG.
+
+.. code-block:: python
+
+    import jinja2
+
+    dag = DAG('simple_dag', template_undefined=jinja2.Undefined)
```
It is also possible to override it on a case-by-case basis by using the `| default` Jinja filter:

```
'{{ a | default(1) }}'
```

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+    dag = DAG('simple_dag', template_undefined=jinja2.Undefined)
+
+
+**Changes to the KubernetesPodOperator**
```
Should these be headings so they show up in the toc?

```suggestion
Changes to the KubernetesPodOperator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

ashb replied:
I don't know about this -- it might be a bit too noisy with that.

vikramkoka replied:
Yes, I was worried it would get way too noisy, with too many sections. 
I went back and forth on this a bit, can easily changed it to be sub-headings, but ended up keeping it like this for now.

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+The DAG-level permission actions, ``can_dag_read`` and ``can_dag_edit`` are going away. They are being replaced with ``can_read`` and ``can_edit``. When a role is given DAG-level access, the resource name (or "view menu", in Flask App-Builder parlance) will now be prefixed with ``DAG:``. So the action ``can_dag_read`` on ``example_dag_id``, is now represented as ``can_read`` on ``DAG:example_dag_ ...
+
+*As part of running ``db upgrade``, existing permissions will be migrated for you.*
```
XD updated this in a PR just "now" -- could you get the lastest version please.

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+Before upgrading to this release, we recommend activating the new FAB RBAC UI. For that, you should set
+the ``rbac`` options  in ``[webserver]`` in the ``airflow.cfg`` file to ``true``
+
+.. code-block:: bash
```
```suggestion
.. code-block:: ini
```

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+You don't need to make changes to the configuration file as the FAB RBAC UI is
+the only supported UI.
+
+.. code-block:: bash
+
+    airflow users create \
+        --role Admin \
+        --username admin \
+        --firstname FIRST_NAME \
+        --lastname LAST_NAME \
+        --email EMAIL@example.org
+
```
```suggestion
```

(Just codifying Jarek's suggestion)

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+
+The following ``airflow.cfg`` values will be deprecated:
+
+.. code-block:: bash
```
```suggestion
The following ``airflow.cfg`` values will be deprecated::
```

This one is not bash -- and no special syntax is appropriate here.

ashb replied:
(Ending a block with `::` makes the next block a code block iff it is indented.)

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+* Pause all the DAGs and make sure there is nothing actively running
+* Upgrade the Airflow version to the 2.0 version of choice
+* Upgrade the Airflow meta database using ``airflow db upgrade``. This command is now shown in the Airflow 2.0 CLI syntax.
+* Restart Airflow Scheduler, Webserver, and Workers
```
I think we should explicitly mention something about installing the right providers, either via the extras or selecting the (non backport) providers too.

Possibly have to _uninstall_ the backport providers too if you are upgrading an install in place (i.e. if you are not using Docker where a change to requirements gets a fresh set of modules)

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1164 @@
+Appendix
+''''''''
+
+Changed Parameters for the KubernetesPodOperator
```
These headings are at the same level -- they probably shouldn't be.

vikramkoka replied:
Changed now to be sub-headings.

vikramkoka committed:
Apply suggestions from code review

Feedback from review

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

vikramkoka committed:
Fixed spelling errors
Fixed spelling errors caused by having missed enclosing some code references
with approriate apostrophes

kaxil committed:
Move Upgrade Check docs to apache-airflow dir

vikramkoka committed:
Incorporated feedback
Changed based on feedback from reviewers.

vikramkoka committed:
Apply suggestions from code review

Incorporated feedback from reviewers

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

vikramkoka committed:
Updated based on feedback

Removed the Upgrading to Airflow 2.0.md file since it is now obsolete
based on this document.
Added sub-headings to the Appendix section for better structure.
Updated the dag permissions changes section.

vikramkoka committed:
Updated based on feedback from review
Specifically around provider installation and Jinja templates.

ashb commented on docs/apache-airflow/upgrading-to-2.rst:
```diff
@@ -0,0 +1,1187 @@
+upgrading the Airflow version.
+
+At this point, just follow the standard Airflow version upgrade process:
+* Make sure your Airflow meta database is backed up
```
```suggestion
At this point, just follow the standard Airflow version upgrade process:

* Make sure your Airflow meta database is backed up
```

ashb committed:
Update docs/apache-airflow/upgrading-to-2.rst

ashb committed:
Fix rst errors
